File: mobile/src/wirelessGecko/java/com/siliconlabs/bledemo/activity/RangeTestModeDialog.java
package com.siliconlabs.bledemo.activity;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import butterknife.ButterKnife;
import butterknife.InjectView;
import butterknife.OnClick;
import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.rangetest.RangeTestMode;
import com.siliconlabs.bledemo.rangetest.TxPower;

import java.text.DecimalFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;

/**
 * @author Comarch S.A.
 */

public class RangeTestModeDialog extends DialogFragment implements RangeTestPresenter.RangeTestView {

    RangeTestPresenter.Controller controller;

    @InjectView(R.id.range_mode_device_name)
    TextView deviceNameView;

    @InjectView(R.id.range_mode_device_number)
    TextView deviceNumberView;

    @InjectView(R.id.range_mode_tx_power)
    TextView powerView;

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        return new Dialog(getActivity(), getTheme()) {
            @Override
            public void onBackPressed() {
                controller.cancelTestMode();
                dismiss();
            }
        };
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setRetainInstance(true);
        setStyle(STYLE_NO_TITLE, getTheme());
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_range_test_mode, container, false);
        ButterKnife.inject(this, view);

        controller = (RangeTestActivity) getActivity();
        controller.setView(this);

        setCancelable(false);

        return view;
    }

    @OnClick(R.id.range_mode_tx)
    void onTxModeButtonClicked() {
        RangeTestPresenter.Controller controller = (RangeTestActivity) getActivity();
        controller.setView(null);

        controller.initTestMode(RangeTestMode.Tx);
        dismiss();
    }

    @OnClick(R.id.range_mode_rx)
    void onRxModeButtonClicked() {
        RangeTestPresenter.Controller controller = (RangeTestActivity) getActivity();
        controller.setView(null);

        controller.initTestMode(RangeTestMode.Rx);
        dismiss();
    }

    @OnClick(R.id.range_mode_cancel)
    void onCancelButtonClicked() {
        controller.cancelTestMode();
        dismiss();
    }

    @Override
    public void runOnUiThread(Runnable runnable) {
        View view = getView();
        if (view != null) {
            view.post(runnable);
        }
    }

    @Override
    public void showDeviceName(String name) {
        deviceNameView.setText(name);
    }

    @Override
    public void showModelNumber(String number) {
        deviceNumberView.setText(number);
    }

    @Override
    public void showTxPower(TxPower power, List<TxPower> values) {
        float value = power.asDisplayValue();
        DecimalFormat formatter = new DecimalFormat("#.##");
        powerView.setText(String.format(Locale.ROOT, "%sdBm", formatter.format(value)));
    }

    @Override
    public void showPayloadLength(int length, List<Integer> values) {
        // not available in this view
    }

    @Override
    public void showMaWindowSize(int size, List<Integer> values) {
        // not available in this view
    }

    @Override
    public void showChannelNumber(int number) {
        // not available in this view
    }

    @Override
    public void showPacketCountRepeat(boolean enabled) {
        // not available in this view
    }

    @Override
    public void showPacketRequired(int required) {
        // not available in this view
    }

    @Override
    public void showPacketSent(int send) {
        // not available in this view
    }

    @Override
    public void showPer(float per) {
        // not available in this view
    }

    @Override
    public void showMa(float ma) {
        // not available in this view
    }

    @Override
    public void showRemoteId(int id) {
        // not available in this view
    }

    @Override
    public void showSelfId(int id) {
        // not available in this view
    }

    @Override
    public void showUartLogEnabled(boolean enabled) {
        // not available in this view
    }

    @Override
    public void showRunningState(boolean running) {
        // not available in this view
    }

    @Override
    public void showTestRssi(int rssi) {
        // not available in this view
    }

    @Override
    public void showTestRx(int received, int required) {
        // not available in this view
    }

    @Override
    public void clearTestResults() {
        // not available in this view
    }

    @Override
    public void showPhy(int phy, LinkedHashMap<Integer, String> values) {
        // not available in this view
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/RangeTest/Fragments/RangeTestModeDialog.kt
package com.siliconlabs.bledemo.RangeTest.Fragments

import android.app.Dialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import com.siliconlabs.bledemo.R
import com.siliconlabs.bledemo.RangeTest.Activities.RangeTestActivity
import com.siliconlabs.bledemo.RangeTest.Models.RangeTestMode
import com.siliconlabs.bledemo.RangeTest.Models.TxPower
import com.siliconlabs.bledemo.RangeTest.Presenters.RangeTestPresenter.Controller
import com.siliconlabs.bledemo.RangeTest.Presenters.RangeTestPresenter.RangeTestView
import kotlinx.android.synthetic.main.dialog_range_test_mode.*
import java.text.DecimalFormat
import java.util.*

/**
 * @author Comarch S.A.
 */
class RangeTestModeDialog : DialogFragment(), RangeTestView {
    lateinit var controller: Controller

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return object : Dialog(requireContext(), theme) {
            override fun onBackPressed() {
                controller.cancelTestMode()
                dismiss()
            }
        }.apply {
            window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setStyle(STYLE_NO_TITLE, theme)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.dialog_range_test_mode, container, false)
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        controller = activity as RangeTestActivity
        controller.setView(this)
        isCancelable = false

        tx_mode.setOnClickListener {
            onTxModeButtonClicked()
        }

        rx_mode.setOnClickListener {
            onRxModeButtonClicked()
        }

        btn_cancel.setOnClickListener {
            onCancelButtonClicked()
        }

    }

    private fun onTxModeButtonClicked() {
        dismiss()
        controller.setView(null)
        controller.initTestMode(RangeTestMode.Tx)
    }

    private fun onRxModeButtonClicked() {
        dismiss()
        controller.setView(null)
        controller.initTestMode(RangeTestMode.Rx)
    }

    private fun onCancelButtonClicked() {
        dismiss()
        controller.cancelTestMode()
    }

    override fun runOnUiThread(runnable: Runnable?) {
        val view = view
        view?.post(runnable)
    }

    override fun showDeviceName(name: String?) {
        tv_device_name.text = name
    }

    override fun showModelNumber(number: String?, running: Boolean?) {
        tv_device_number.text = number
    }

    override fun showTxPower(power: TxPower?, values: List<TxPower>) {
        val value = power?.asDisplayValue()
        val formatter = DecimalFormat("#.##")
        tv_tx_power.text = String.format(Locale.ROOT, "%sdBm", formatter.format(value?.toDouble()))
    }

    override fun showPayloadLength(length: Int, values: List<Int>) {
        // not available in this view
    }

    override fun showMaWindowSize(size: Int, values: List<Int>) {
        // not available in this view
    }

    override fun showChannelNumber(number: Int) {
        // not available in this view
    }

    override fun showPacketCountRepeat(enabled: Boolean) {
        // not available in this view
    }

    override fun showPacketRequired(required: Int) {
        // not available in this view
    }

    override fun showPacketSent(sent: Int) {
        // not available in this view
    }

    override fun showPer(per: Float) {
        // not available in this view
    }

    override fun showMa(ma: Float) {
        // not available in this view
    }

    override fun showRemoteId(id: Int) {
        // not available in this view
    }

    override fun showSelfId(id: Int) {
        // not available in this view
    }

    override fun showUartLogEnabled(enabled: Boolean) {
        // not available in this view
    }

    override fun showRunningState(running: Boolean) {
        // not available in this view
    }

    override fun showTestRssi(rssi: Int) {
        // not available in this view
    }

    override fun showTestRx(received: Int, required: Int) {
        // not available in this view
    }

    override fun clearTestResults() {
        // not available in this view
    }

    override fun showPhy(phy: Int, values: LinkedHashMap<Int, String>) {
        // not available in this view
    }
}